File: packages/typia/src/programmers/internal/check_template.ts
import ts from "typescript";

import { MetadataTemplate } from "../../schemas/metadata/MetadataTemplate";

import { ICheckEntry } from "../helpers/ICheckEntry";
import { template_to_pattern } from "./template_to_pattern";

/** @internal */
export const check_template = (props: {
  templates: MetadataTemplate[];
  input: ts.Expression;
}): ICheckEntry => {
  // TYPEOF STRING & TAGS
  const conditions: ts.Expression[] = [
    ts.factory.createStrictEquality(
      ts.factory.createStringLiteral("string"),
      ts.factory.createTypeOfExpression(props.input),
    ),
  ];

  // TEMPLATES
  const internal: ts.Expression[] = props.templates.map((tpl) =>
    ts.factory.createCallExpression(
      ts.factory.createIdentifier(
        `RegExp(/${template_to_pattern({
          top: true,
          template: tpl.row,
        })}/).test`,
      ),
      undefined,
      [props.input],
    ),
  );
  conditions.push(
    internal.length === 1
      ? internal[0]!
      : internal.reduce((x, y) => ts.factory.createLogicalOr(x, y)),
  );

  // COMBINATION
  return {
    expression: conditions.reduce((x, y) => ts.factory.createLogicalAnd(x, y)),
    conditions: [],
    expected: props.templates.map((tpl) => tpl.getName()).join(" | "),
  };
};

File: packages/core/src/programmers/iterate/check_template.go
package iterate

import h "github.com/samchon/typia/packages/core/src/programmers/helpers"

func check_template(input h.Expression, template h.Metadata) h.ICheckEntry {
	pattern := h.PatternFix(template_to_pattern(template))
	expr := h.Expr("new RegExp(" + h.QuoteString(pattern) + ").test(" + string(input) + ")")
	return h.ICheckEntry{Expected: "`" + pattern + "`", Expression: &expr}
}
